### PR TITLE
Immutable typed-list

### PR DESCRIPTION
**Pull request by esc**
This is the runtime approach to the immutable typed-list. An additional member is added
to the underlying C-Struct and exposed to the compiler and Python levels above
as usual.

### CODE REVIEW

**Comment by esc** [comment]
This may replace #5201 if the approach is deemed acceptable.

File: numba/tests/test_listimpl.py [hunk 1]
```diff
@@ -421,3 +458,39 @@ def test_sizing(self):
...
+        self.assertTrue(l.is_mutable)
+        self.assertEqual(len(l), 1)
+        r = struct.unpack("q", l[0])[0]
+        self.assertEqual(r, 1)
+
+        # set to immutable and test guards
+        l.set_immutable()
+        self.assertFalse(l.is_mutable)
+        # append
+        with self.assertRaises(ValueError):
```
Comment (sklam) [inline numba/tests/test_listimpl.py]: Can you make this check for the error message? `ValueError` is common. This can be caused by other reasons.

Comment (esc) [reply numba/tests/test_listimpl.py]: done

**Comment by esc** [comment]
@sklam thanks for the review, I have addressed your comments.

**Pull request merged**